### ISSUE TITLE
Functions to be added

### ISSUE THREAD

**Issue by richardsc** [TEOS-10/GSW-R#1]
This issue serves to organize and document the adding of GSW functions to the package, by way of following the [developer instructions](https://github.com/TEOS-10/GSW-R/blob/master/README_developer.md). The order is alphabetical (as listed in `src/gswteos-10.h`), but if users have a specific need for a function that has not yet been completed they should comment here.

## Functions not to be added

The following functions are listed in the ``src/gswteos-10.h`` file but they will not be added to the package because they are for internal use by those who are developing the TEOS functions:

* gsw_add_barrier
* gsw_add_mean
* gsw_xinterp1
* gsw_indx
* gsw_gibbs                                  the TEOS-10 Gibbs function of seawater and its derivatives
* gsw_gibbs_ice                              the TEOS-10 Gibbs function of ice and its derivatives
* gsw_SAAR                                   Absolute Salinity Anomaly Ratio (excluding the Baltic Sea)
* gsw_Fdelta                                 ratio of Absolute to Preformed Salinity, minus 1
* gsw_deltaSA_atlas                          Absolute Salinity Anomaly atlas value (excluding the Baltic Sea)
* gsw_SA_from_SP_Baltic                      Calculates Absolute Salinity in the Baltic Sea
* gsw_SP_from_SA_Baltic                      Calculates Practical Salinity in the Baltic Sea
* gsw_infunnel                               "oceanographic funnel" check for the 48-term equation
* gsw_entropy_part                           entropy minus the terms that are a function of only SA
* gsw_entropy_part_zerop                     entropy_part evaluated at 0 dbar
* gsw_interp_ref_cast                        linearly interpolates the reference cast
* gsw_interp_SA_CT                           linearly interpolates (SA,CT,p) to the desired p
* gsw_gibbs_pt0_pt0                          gibbs(0,2,0,SA,t,0)
* gsw_gibbs_ice_part_t                       part of gibbs_ice(1,0,t,p)
* gsw_gibbs_ice_pt0                          part of gibbs_ice(1,0,pt0,0)
* gsw_specvol_SSO_0_p                        specvol_CT(35.16504,0,p)
* gsw_enthalpy_SSO_0_p                       enthalpy_CT(35.16504,0,p)
* gsw_Hill_ratio_at_SP2

## Functions to be added

As functions are added (meaning coded, documented and tested), their tick boxes are checked off. To avoid edit conflicts, the developers can communicate by means of comments in this thread (e.g. "In the next 2 hours I intend to code such-and-such"), and these comments are erased when the code is pushed back to github. Thus, the idea is that a developer starts by reloading this webpage, picks a task that is of interest, adds a comment, and usually an hour later pushs the repository and deletes the comment.

A link providing docs for the functions is [http://www.teos-10.org/pubs/gsw/html/gsw_contents.html](http://www.teos-10.org/pubs/gsw/html/gsw_contents.html).

* [x] gsw_adiabatic_lapse_rate_from_ct
* [x] gsw_alpha
* [x] gsw_alpha_on_beta
* [x] gsw_alpha_wrt_t_exact
* [x] gsw_beta_const_t_exact
* [x] gsw_beta
* [x] gsw_c_from_sp
* [x] gsw_cabbeling
* [x] gsw_ct_freezing **done, but yields check values from `gsw_CT_freezing_poly`**
* [x] gsw_ct_from_pt
* [x] gsw_ct_from_t
* [x] gsw_deltasa_from_sp
* [x] gsw_dynamic_enthalpy
* [x] gsw_enthalpy
* [x] gsw_enthalpy_t_exact
* [x] gsw_cp_t_exact
* [x] gsw_entropy_from_t
* [x] gsw_grav
* [x] gsw_internal_energy
* [x] gsw_ipv_vs_fnsquared_ratio
* [x] gsw_kappa
* [x] gsw_kappa_t_exact
* [x] gsw_latentheat_evap_ct
* [x] gsw_latentheat_evap_t
* [x] gsw_latentheat_melting
* [x] gsw_nsquared
* [x] gsw_pot_rho_t_exact
* [x] gsw_pt0_from_t
* [x] gsw_pt_from_ct
* [x] gsw_pt_from_t
* [x] gsw_rho
* [x] gsw_rho_first_derivatives
* [x] gsw_rho_t_exact
* [x] gsw_sa_from_rho
* [x] gsw_sa_from_sp
* [x] gsw_sa_from_sstar
* [x] gsw_sigma0
* [x] gsw_sigma1
* [x] gsw_sigma2
* [x] gsw_sigma3
* [x] gsw_sigma4
* [x] gsw_sound_speed
* [x] gsw_sound_speed_t_exact
* [x] gsw_specvol_anom
* [x] gsw_specvol
* [x] gsw_specvol_t_exact
* [x] gsw_sp_from_c
* [x] gsw_sp_from_sa
* [x] gsw_sp_from_sk
* [x] gsw_sp_from_sr
* [x] gsw_sp_from_sstar
* [x] gsw_sr_from_sp
* [x] gsw_sstar_from_sa
* [x] gsw_sstar_from_sp
* [x] gsw_t_freezing **done, but yields check values from `gsw_t_freezing_poly`**
* [x] gsw_t_from_ct
* [x] gsw_thermobaric
* [x] gsw_turner_rsubrho
* [x] gsw_z_from_p

**Comment by dankelley** [comment]
I am going to take 10 minutes and put the functions in alphabetical order.

**Comment by dankelley** [comment]
Through 4 days of use, the list of desired functions has not expanded, so I'm closing this. I think we're done ... of course, pending possible updates if the C library changes.

**Comment by fdelahoyde** [comment]
Hi Dan,

GSW-C is now on github with my latest changes.

Frank

**Comment by dankelley** [comment]
Thanks, Frank, this is wonderful.  I'll incorporate it into the R version sometime today.

A tiny thing: is there a reason why you use ``#include <gswteos-10.h>`` instead of ``#include "gswteos-10.h"``? 

The reason I ask is that the latter should still work properly in your builds, but it will also work properly in my builds within R.  For R packages to get to users, they are built up on independent systems over which the user has no control, and I don't think the systems permit installing .h files in ``/usr/...`` or whatever, or altering the value of the ``PATH``.  This is a tiny thing of course; obvsiously it took me 10 times long to post this than it would take to tweak the source.

**Comment by fdelahoyde** [comment]
Yes, I use the "include <gswteos-10.h>" syntax and add "-I." to CFLAGS in the
Makefile. This is because one purpose of this package for me is to build a
binary RPM for RedHat/CentOS systems that does indeed install into /usr.
